File: test/four/app/src/main/java/com/z3d/material/MaterialBaseParam.java
package com.z3d.material;

import com.z3d.base.Scene_data;
import com.z3d.base.TexTuresBackFun;
import com.z3d.engine.GC;
import com.z3d.filemodel.TextureManager;
import com.z3d.res.MaterialInfoVo;
import com.z3d.scene.Scene3D;

import java.util.ArrayList;
import java.util.List;

public class MaterialBaseParam extends GC {
    public Material material;
    public List<DynamicBaseTexItem> dynamicTexList ;
    public List<DynamicBaseConstItem> dynamicConstList ;

    public Scene3D scene3D;
    public MaterialBaseParam(Scene3D val){
        this.scene3D=val;
    }
    public void setData(Material mater, List<MaterialInfoVo> ary )
    {
        this.material = mater;
        this.dynamicConstList = new ArrayList();
        this.dynamicTexList = new ArrayList<>();
        List<ConstItem> constList  = mater.constList;
        List<TexItem> texList  = mater.texList;
        for (int i = 0; i < ary.size(); i++) {
            try {

                MaterialInfoVo obj = (MaterialInfoVo) ary.get(i);

                int objType= (int) obj.type;
                if (objType== 0) {
                final    DynamicBaseTexItem texItem = new DynamicBaseTexItem(scene3D);
                    texItem.paramName = (String) obj.name;
                    for (int j = 0; j < texList.size(); j++) {
                        if (texItem.paramName.equals(texList.get(j).paramName)) {
                            texItem.target = texList.get(j);
                            break;
                        }
                    }
                    int mipmap = 0;
                    if (texItem.target!=null) {
                        mipmap = texItem.target.mipmap;
                    }
                    mipmap = 0;
                    this.scene3D.textureManager.getTexture(Scene_data.fileRoot+ obj.url, new TexTuresBackFun() {
                        @Override
                        public void Bfun(TextureRes value) {
                            texItem.textureRes = value;
                        }
                    });
                    this.dynamicTexList.add(texItem);
                } else {
                    String targetName = (String)  obj.name;
                    ConstItem target=null ;
                    for (int j = 0; j < constList.size(); j++) {
                        if (targetName.equals( constList.get(j).paramName0)
                                || targetName.equals( constList.get(j).paramName1)
                                || targetName.equals(constList.get(j).paramName2)
                                || targetName .equals(constList.get(j).paramName3)) {

                            target = constList.get(j);
                            break;

                        }
                    }
                    DynamicBaseConstItem constItem = new DynamicBaseConstItem();
                    constItem.setTargetInfo(target, targetName, (int)objType);
                    List<Float> valArr=new ArrayList<>();
                    if (objType>= 1) {
                        valArr.add((float) obj.x);
                    }
                    if (objType >= 2) {
                        valArr.add((float) obj.y);
                    }
                    if (objType >= 3) {
                        valArr.add((float) obj.z);
                    }
                    constItem.setCurrentVal(valArr);
                    this.dynamicConstList.add(constItem);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void update() {
        if (this.material!=null && this.dynamicConstList!=null) {
            for (int i = 0; i < this.dynamicConstList.size(); i++) {
                this.dynamicConstList.get(i).update();
            }
        }
    }
}

File: listios/StorageStarted/pan3d/material/MaterialBaseParam.m
//
//  MaterialBaseParam.m
//  iosgl
//
//  Created by zhao on 9/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "MaterialBaseParam.h"
#import "DynamicBaseConstItem.h"
#import "ConstItem.h"
#import "TexItem.h"
#import "Scene_data.h"
#import "TextureManager.h"

@implementation MaterialBaseParam
-(void)setData:(Material*)material ary:(NSArray<NSDictionary*>*)ary;
{
    MaterialBaseParam* this=self;
    this.material = material;
    this.dynamicConstList =  [[NSMutableArray alloc]init];
    this.dynamicTexList = [[NSMutableArray alloc]init];
    
    NSMutableArray<ConstItem*>* constList  = material.constList;
    NSMutableArray<TexItem*>* texList  = material.texList;
    
    for (int i = 0; i < ary.count; i++) {
        
        NSDictionary* obj = ary[i];
        int obj_type=[obj[@"type"]intValue];
        NSString* obj_name=(NSString*)obj[@"name"] ;
        if (obj_type == 0) {
            DynamicBaseTexItem* texItem   = [[DynamicBaseTexItem alloc]init:self.scene3D];
            texItem.paramName = obj_name;
            for (int j = 0; j < texList.count; j++) {
                if (texItem.paramName == texList[j].paramName) {
                    texItem.target = texList[j];
                    break;
                }
            }
            int mipmap = 0;
            if (texItem.target) {
                mipmap = texItem.target.mipmap;
            }
            mipmap = 0;
            
            [self.scene3D.textureManager getTexture:[[Scene_data default]getWorkUrlByFilePath:obj[@"url"]] fun:^(NSObject * _Nonnull any) {
                texItem.textureRes=(TextureRes*)any;
            } wrapType:0 info:nil filteType:0 mipmapType:0];
            
             [this.dynamicTexList addObject:texItem];
            
            
        } else {
            NSString* targetName = obj_name;
            ConstItem* target = nil;
            for (int j = 0; j < constList.count; j++) {
                if (targetName == constList[j].paramName0
                    || targetName == constList[j].paramName1
                    || targetName == constList[j].paramName2
                    || targetName == constList[j].paramName3) {
                    target = constList[j];
                    break;
                }
            }
            DynamicBaseConstItem* constItem = [[DynamicBaseConstItem alloc]init];
            [constItem setTargetInfo:target paramName:targetName type: obj_type ];
            if (obj_type== 1) {
                [constItem setCurrentVal:obj[@"x"]];
            } else if (obj_type== 2) {
                [constItem setCurrentVal:obj[@"x"] y:obj[@"y"]];
            } else {
                [constItem setCurrentVal:obj[@"x" ] y:obj[@"y"] z:obj[@"z"]];
            }
            [this.dynamicConstList addObject:constItem];
            
            
        }
    }
    
}

-(void)update;
{
    //[0]    DynamicBaseConstItem *    0x2838371e0    0x00000002838371e0
     MaterialBaseParam* this=self;
    if (this.material && this.dynamicConstList) {
        for (int i = 0; i < this.dynamicConstList.count; i++) {
            [this.dynamicConstList[i] update:0];
        }
    }
}
/*
 public update(): void {
        if (this.material && this.dynamicConstList) {
            for (var i: number = 0; i < this.dynamicConstList.length; i++) {
                this.dynamicConstList[i].update();
            }
        }
    }
 */
@end
